wallet-new/src/Cardano/Wallet/Kernel/DB/Sqlite.hs

```diff
@@ -0,0 +1,562 @@
+
+        metaQuery = case mbSorting of
+            Nothing  -> SQL.all_ (_mDbMeta metaDB)
+            Just (Sorting SortByCreationAt dir) ->
+                SQL.orderBy_ (toBeamSortDirection dir . _txMetaTableCreatedAt) $ SQL.all_ (_mDbMeta metaDB)
+            Just (Sorting SortByAmount     dir) ->
+                SQL.orderBy_ (toBeamSortDirection dir . _txMetaTableAmount) $ SQL.all_ (_mDbMeta metaDB)
+
+        -- The following two queries are disjointed and both fetches, respectively,
+        -- a list of tuples of type @(TxMeta, TxInput)@ and @(TxMeta, TxOutput)@.
+        -- The rationale behind doing two separate queries is that there is no elegant
+        -- way to express a 3-table join without having 'Maybe's cropping up in the
+        -- Haskell result sets, and furthermore we would have to deal with duplicates.
+        -- Doing two separate queries is yes more I/O taxing, but spares us from doing
+        -- duplicate filtering on the Haskell side.
+        paginatedInputs = SQL.select $ do
+            meta   <- SQL.limit_ limit (SQL.offset_ offset metaQuery)
+            inputs  <- SQL.oneToMany_ (_mDbInputs metaDB)  (getTx _getTxInput) meta
+            pure (meta, inputs)
+
+        paginatedOutputs = SQL.select $ do
+            meta   <- SQL.limit_ limit (SQL.offset_ offset metaQuery)
+            outputs <- SQL.oneToMany_ (_mDbOutputs metaDB) (getTx _getTxOutput) meta
+            pure (meta, outputs)
+
+        -- | Groups the inputs or the outputs under the same 'TxMeta'.
+        transform :: NonEmpty (TxMeta, a) -> M.Map OrdByCreationDate (NonEmpty a)
+        transform = Foldable.foldl' updateFn M.empty
+
+        updateFn :: M.Map OrdByCreationDate (NonEmpty a)
+                 -> (TxMeta, a)
+                 -> M.Map OrdByCreationDate (NonEmpty a)
+        updateFn acc (txMeta, new) =
+            M.insertWith (<>) (OrdByCreationDate txMeta) (new :| []) acc
+
+        toValidKernelTxMeta :: M.Map OrdByCreationDate (NonEmpty TxInput)
+                            -> [Kernel.TxMeta]
+                            -> (OrdByCreationDate, NonEmpty TxOutput)
+                            -> [Kernel.TxMeta]
+        toValidKernelTxMeta inputMap acc (OrdByCreationDate t, outputs) =
+            case M.lookup (OrdByCreationDate t) inputMap of
+                 Nothing     -> acc
+                 Just inputs -> toTxMeta t inputs outputs : acc
+
+-- | Generates a Beam's AST fragment for use within a SQL query, to order
+-- the results of a @SELECT@.
+toBeamSortDirection :: SortDirection
+                    -> SQL.QExpr (Sql92OrderingExpressionSyntax (Sql92SelectOrderingSyntax SqliteSelectSyntax)) s a
+                    -> SQL.QOrd (Sql92SelectOrderingSyntax SqliteSelectSyntax) s a
+toBeamSortDirection Ascending  = SQL.asc_
+toBeamSortDirection Descending = SQL.desc_
+
+-- | Generates a function suitable for 'sortBy' out of a 'Sorting'.
+toOrdering :: Sorting -> (Kernel.TxMeta -> Kernel.TxMeta -> Ordering)
```

Comment:
why the parens?

Reply:
Just a visual cue that this really returns a function, but I don't have strong feelings about it.

Reply:
ah, but everything really returns a function :smile: :sheep: :dancer: